### PR TITLE
Wireless protocol code

### PR DESCRIPTION
**Pull request by Ravenslofty**
This supercedes #74 and #107, because the protocol turns out to have a lot of code in common.

The commits are...very messy, but the code lights up a Dark Core. Next step is the K63 Wireless.

### CODE REVIEW

**Comment by fleischie** [comment]
Hello.

I will have a week of vacation this coming week (25. June - 29. June), access to all them OSs (Win10, MacOS, Arch Linux, Raspbian, etc.) aaand a Corsair Dark Core.

Do you still need help with this? What can I do? What would be a good scenario to test?

Looking forward to helping you guys once more. 👍

**Comment by ZirconiumX** [comment]
Holy crap, fleischie, we thought you'd gone entirely, just like l2y.

Mostly it's just a matter of code refactoring, which I haven't gotten round
to.

**Comment by fleischie** [comment]
Nope, I have just gotten exceptionally bad at having free time. But I'd like to spend some vacation time with you. ❤️

**Comment by DewaldDeJager** [comment]
I'm also willing to test this.

**Comment by DewaldDeJager** [comment]
@ZirconiumX When building the wireless branch and opening the GUI, I just get the "No devices connected" message. My version is `0.3.1-42-g38328b1`.

**Comment by tatokis** [comment]
Try plugging it in with a USB wire and not wirelessly.

**Comment by gilvbp** [comment]
How can i help?

**Comment by ZirconiumX** [comment]
@gilvbp We need to mostly refactor this code to add wireless keyboard and mouse vtables.

**Comment by gilvbp** [comment]
@ZirconiumX i will test the dark core and k63wl on wired mode thanks.

**Comment by rodgomesc** [comment]
Why has not this been merged?,

**Comment by ZirconiumX** [comment]
Lack of testing and lack of motivation.

**Comment by gilvbp** [comment]
I have tested, dark core work on usb mode. But freezes at wifi mode after some seconds.

**Comment by ZirconiumX** [comment]
And that's why this isn't merged.

**Comment by rodgomesc** [comment]
any progress in wireless branch?

**Comment by ZirconiumX** [comment]
Not really.

**Comment by rodgomesc** [comment]
What exactly did not work on this mouse the last time you tested it? I'm using fedora 30 silverblue, and it's working perfectly without installing any additional software (maybe it's the newer version of the kernel), to control the lighting I wrote a little [ script ](https://gist.github.com/rodgomesc/d217815ed98fe14d057dd6293a51dd53)in python (language I'm more familiar with) for tests and it also worked normally

**Comment by tatokis** [comment]
Yes. We have that working too. 

The problems are handling input when both wireless and wired are plugged in, and querying things like battery.

Additionally software animations aren't supported, and I don't want to abuse the static colour profile to mimic animations.

If you just want to set a static colour profile, the wireless branch will do that. 

I don't know if it can set DPI.

**Comment by lss4** [comment]
I just tried the f683f13ef89ced39522c2c61810cf9e3dd38ba89 commit by installing from the AUR, using ckb-next-git PKGBUILD but edited to use the wireless branch.

It builds and could detect my Dark Core SE, but the mouse no longer functions properly. Moving the mouse no longer moves the cursor but spits some random keyboard inputs like the following (I'm using zsh):
```
uuuuuuu-[;',. ]\                                                                                                                                                                                                                   127 ↵
quote> 1234567890wyuzxv-[;',. ]\
> 1234567890wyuzxv-[;',. ]\
quote> 1234567890wyuzxv-[;',. ]\
> 1234567890wyuzxv-[;',. ]\
quote> -[;',. ]\
> =1234567890wyuzxv-[;',. ]\
quote>  
quote> 1234567890wyuzxv1234567890wyuzxv-[;',. ]\
> 1234567890wyuzxv1234567890wyuzxv1234567890wyuzx[;'`1234567890wyuzxv
quote> 1234567890wyuzxv1234567890wyuzxv
quote> 1234567890wyuzxv
quote> 1234567890wyuzxv1234567890wyuzxv
quote> 1234567890wyuzxv
quote> 
quote> 234567890wyuzxv1234567890wyuzxv
quote> 1234567890wyuzxv
quote> 1234567890wyuzxv
quote> 1234567890wyuzxv
quote> 
quote> 
quote> 
```

Not sure what might have caused it. I'm using the USB receiver (wireless mode) anyway.

**Comment by tatokis** [comment]
Should be fixed by https://github.com/ckb-next/ckb-next/commit/8f3ca3185fc7162fb8cfd9272fe487af68dcfc2d

**Comment by beniwtv** [comment]
I can confirm it works now for me after that last commit :)
The only thing that doesn't is saving to the hardware, not sure if that is even supported on a Dark Core SE tho.

**Comment by lss4** [comment]
> I can confirm it works now for me after that last commit :)
> The only thing that doesn't is saving to the hardware, not sure if that is even supported on a Dark Core SE tho.

Yeah, it's working for me, too.

You need to connect the USB cable in order to save the profile to the hardware (which was the case for iCUE on Windows, haven't tested here yet).

It seems for now I can't get breathing light working like it was on Windows.

**Comment by tatokis** [comment]
> Additionally software animations aren't supported, and I don't want to abuse the static colour profile to mimic animations. If you just want to set a static colour profile, the wireless branch will do that.

**Comment by lss4** [comment]
> > Additionally software animations aren't supported, and I don't want to abuse the static colour profile to mimic animations. If you just want to set a static colour profile, the wireless branch will do that.

I remember I can save breathing light (Color Pulse) effects to the hardware with iCUE on Windows, and it'll also work on other systems that don't have iCUE installed.

For now, getting light effects working might be a lower priority, if there are still things needed to be done in order to get the core functionalities of the mouse working here without issues.

**Comment by beniwtv** [comment]
I tried with a cable, but the save to hardware button remains greyed out

**Comment by lss4** [comment]
Another issue...

It seems Dark Core SE mouse itself (connected via USB) and Dark Core SE's wireless USB receiver are treated as separate devices, and thus they cannot share the same configurations like in iCUE.

**Comment by OrsanZ** [comment]
Does the Dark Core work now?
I tried building the wireless branch and it failed, so I couldn't test it.
If not, I've recently rewrote my "personal" utility for the dark core and am now feeling confident enough to get to the real deal and implement it here, since the branch hasn't been updated in half a year.

Do I just start by making a fork?

**Comment by Ravenslofty** [comment]
The answer is that we don't know if the Dark Core works because testers are
few and far between.

**Comment by OrsanZ** [comment]
Oh, okay.
I mean I have an SE, so I could help with that.
Sadly I can't get the wireless branch to compile, is that an error that has to be fixed first or am I doing something wrong?
The error I get is `multiple definition of `devpath'`

**Comment by Ravenslofty** [comment]
That's odd. Did you make sure to run `make clean`?

(It's almost midnight local time here; could you poke me again tomorrow
morning?)

**Comment by tatokis** [comment]
You're probably using the latest gcc. Rebase the branch to master and it'll compile

**Comment by OrsanZ** [comment]
That's probably it. Rebasing seems to be a bit more complicated because of conflicts, so I'll deal with that tomorrow and report my results.

**Comment by OrsanZ** [comment]
Okay, I managed to get it to run after rebasing, as expected.
Overall, most of the features actually work. The only things that break are those, which aren't supported in iCUE, like for example 6 DPI profiles. So here's the report.

### Performance

#### DPI
The Dark Core only supports 4 DPI profiles (3 + sniper) in iCUE, and probably also in firmware. Because of that, switching to and from profile 4/5 causes the Dark Core to get DPI values wrong. Also, the DPI indicator doesn't work for me. The solution to most, or maybe all of this may be to just disable profile 4 and 5 for the Dark Core. Independent XY values also just work.

#### Lift Height
The same type of issue occurs when adjusting the lift height. ckb-next allows the user to choose from more options than the Dark Core supports. The Dark Core only supports low, medium and high, which correspond internally to 03, 04 and 05. This should also be an easy fix. Sadly, there is no concrete way for me to check if I'm right here, since it's not objective, but I do notice the difference when using my own program to change my lift height.

#### Angle Snapping
I also have no idea if angle snapping works, since I don't notice the difference.

### Bindings and Macros
They work just as expected. Nice.

### Device Tab
The device properties (firmware version and serial number) are displayed correctly, but you can choose a keyboard layout for the mouse, which is weird, but doesn't seem to break anything. Adding the battery status and dongle polling rate would be fitting here. There is an empty line in wireless, where the poll rate is in wired. This might be because the dongle polling rate can actually be changed, but my mouse may be broken.

### Lighting
Static colors work fine, but animations only work in wired mode. The lights on the device change seemingly at random in wireless, which is probably due to latency. This likely is also why Corsair doesn't support lighting link on this device.  Sadly, I believe that we have to follow Corsair's lazy workaround of limiting animations. We could still allow custom animations in wired, since wired and wireless count as separate devices and don't share configs anyway.

### TLDR
- Dark Core in wireless should only support firmware animations
- Remove DPI profile 4 and 5
- Remove medium-low and medium-high from lift height options
- DPI indicators are broken
- Battery status would be nice
- Remove the keyboard layout option from _Device_ tab

Hope this helps :slightly_smiling_face:

**Comment by tatokis** [comment]
> Remove the keyboard layout option from Device tab

This is actually intentional, because when you create a macro to type something, it follows the layout setting.
In the past we forced one specific layout, but that created issues depending on the environment. I don't really remember the details.

> Battery status would be nice

You'll need to reverse engineer that yourself and tell us how it's implemented. CUE didn't seem to poll the device and ask it for the battery when I used the emulated dark core, so I have no way of figuring it out.

> Dark Core in wireless should only support firmware animations

In general we need to link devices together and classify them as dongles or wired. This will be implemented for bragi first most likely.

**Comment by OrsanZ** [comment]
> You'll need to reverse engineer that yourself and tell us how it's implemented

As luck would have it, I already did :wink: 

Taken from [here](https://github.com/OrsanZ/OpenDarkCore/blob/master/docs/protocol.md)

#### 0e 50

**Response: 0e 50 00 00 BA 01**

Reads the battery status.

| Abbreviation | Values                                                     |
| ------------ | --------------------------------------------------------- |
| BA           | 00 Not connected, 01 Critical, 02 Low, 03 Medium, 04 High |

> Also, regarding angle snapping, it's easy to tell if it works.
> Open up gimp and try to draw a horizontal line. It's much easier to do with angle snapping enabled.

Thanks, that means it works.

**Comment by Ravenslofty** [comment]
Ooh, this is excellent!

For what it's worth, we have a protocol decoder [here](https://github.com/ckb-next/cueshark) that can decode at least part of `0e ae` (which we think is a radio status packet). If you want, you can add these to the decoder; that would help us a lot.

**Comment by OrsanZ** [comment]
Sure thing. I'll have to figure out how these LUA scripts work, but I'll be sure to add it.

Glad I can help.

**Comment by OrsanZ** [comment]
Okay, I made a pull request. I noticed that you were also missing what I believe to be a query for whether the mouse is connected to the dongle or not.

_0e 51_ returns _0e 00 00 01_ if the device isn't connected and _0e 00 00 02_ when it is. 0e 51 is also omitted if the mouse is connected via cable.

**Comment by tatokis** [comment]
Since I'm at it, here are all the hardware animations that I've had sitting on my computer for a while now.
[nxp_wl_anims.txt](https://github.com/ckb-next/ckb-next/files/5012616/nxp_wl_anims.txt)

**Comment by OrsanZ** [comment]
I'd like to continue by fixing the things, so that the mouse works flawlessly in wired mode. That would includes:

- Remove DPI profile 4 and 5
- Remove medium-low and medium-high from lift height options
- Fix DPI indicators
- (Maybe) Display the battery status

I'd probably be done later today and would make a pull request. Would that be okay?

**Comment by Ravenslofty** [comment]
Yep, pull requests are fine. Heck, you might as well just pull request your code and we can close this PR for that one.

**Comment by OrsanZ** [comment]
Sure. It'd be nice if you could rebase the wireless branch first, since I'm afraid I might break support if I try myself.

**Comment by Ravenslofty** [comment]
I gave it my best shot.

**Comment by tatokis** [comment]
Merged to master indirectly after being bundled with bragi.
